lib/ansible/modules/tacp/tacp_instance.py

```diff
@@ -80,278 +78,713 @@
-        state=dict(type='str', required=True,
-                   choices=STATE_ACTIONS),
-        datacenter=dict(type='str', required=False),
-        migration_zone=dict(type='str', required=False),
-        storage_pool=dict(type='str', required=False),
-        template=dict(type='str', required=False),
-        vcpu_cores=dict(type='int', required=False),
-        memory=dict(type='str', required=False),
-        disks=dict(type='list', required=False),
-        nics=dict(type='list', required=False),
-        vtx_enabled=dict(type='bool', default=True, required=False),
-        auto_recovery_enabled=dict(type='bool', default=True, required=False),
-        description=dict(type='str', required=False),
-        vm_mode=dict(type='str', default='Enhanced', choices=['enhanced', 'Enhanced',
-                                                              'compatibility', 'Compatibility']),
-        application_group={
-            'type': 'str',
-            'required': False,
-        },
+RESULT = {
+    'changed': False,
+    'args': []
+}
+
+MODULE = AnsibleModule(
+    argument_spec=MODULE_ARGS,
+    supports_check_mode=True
+)
+
+# Define configuration
+CONFIGURATION = tacp_utils.get_configuration(MODULE.params['api_key'],
+                                             MODULE.params['portal_url'])
+API_CLIENT = tacp.ApiClient(CONFIGURATION)
+
+RESOURCES = {
+    'app': tacp_utils.ApplicationResource(API_CLIENT),
+    'app_group': tacp_utils.ApplicationGroupResource(API_CLIENT),
+    'datacenter': tacp_utils.DatacenterResource(API_CLIENT),
+    'migration_zone': tacp_utils.MigrationZoneResource(API_CLIENT),
+    'storage_pool': tacp_utils.StoragePoolResource(API_CLIENT),
+    'template': tacp_utils.TemplateResource(API_CLIENT),
+    'update_app': tacp_utils.UpdateApplicationResource(API_CLIENT),
+    'vlan': tacp_utils.VlanResource(API_CLIENT),
+    'vnet': tacp_utils.VnetResource(API_CLIENT)
+}
+
+
+def fail_with_reason(reason):
+    RESULT['msg'] = reason
+    MODULE.fail_json(**RESULT)
+
+
+def fail_and_rollback_instance_creation(reason, instance):
+    RESULT['msg'] = reason + "\n Rolled back application instance creation."
+    RESOURCES['app'].delete(instance.uuid)
+    MODULE.fail_json()
+
+
+def get_parameters_to_create_new_application(playbook_instance):
+    """Given the input configuration for an instance, generate
+        parameters for creating the appropriate payload elsewhere.
 
+    Args:
+        playbook_instance (dict): The specified instance configuration from
+        the playbook input
+
+    Returns:
+        dict: The parameters to be provided to an ApiCreateApplicationPayload
+            object.
+    """
+    parameters_to_create_new_application = {}
+    parameters_to_create_new_application['instance_name'] = playbook_instance['name']  # noqa
+
+    resources = [
+        'datacenter',
+        'migration_zone',
+        'storage_pool',
+        'template'
+    ]
+
+    for resource in resources:
```

Comment:
You are abusing the `resource` naming. 

We already have the resource classes in the utils module, plus the resource instances in the `RESOURCE` mapping here, then having a resource iterator which goes through resources and below compute a resource list just to fetch the resource uuid .. just confusing for me. These names here are used just in this local scope, just in this for loop

I suggest you go a bit more abstract and use disposable names such as  `item` as iterator and place the list directly in the for statement. Also, not really important, but you can use a tuple for less overhead

```suggestion
    for item in ('datacenter', 'migration_zone', 'storage_pool', 'template'):
    
```